chain/rust/src/builders/input_builder.rs

```diff
@@ -0,0 +1,135 @@
+use crate::builders::witness_builder::{InputAggregateWitnessData, PartialPlutusWitness};
+
+use super::witness_builder::{RequiredWitnessSet, NativeScriptWitnessInfo, PlutusScriptWitness};
+
+use crate::{
+    NativeScript,
+    address::Address,
+    certs::StakeCredential,
+    crypto::hash::hash_plutus_data,
+    transaction::{TransactionInput, TransactionOutput, RequiredSigners},
+    plutus::PlutusData,
+};
+
+ #[derive(Debug, thiserror::Error)]
+ pub enum InputBuilderError {
+     #[error("UTXO address was not a payment key: {0:?}")]
+     UTXOAddressNotPayment(Address),
+     #[error("Missing the following witnesses for the input: {0:?}")]
+     MissingWitnesses(RequiredWitnessSet),
+ }
+
+pub fn input_required_wits(utxo_info: &TransactionOutput, required_witnesses: &mut RequiredWitnessSet) {
+    if let Some(cred) = utxo_info.address().payment_cred() {
+        match cred {
+            StakeCredential::PubKey { hash, .. } => {
+                required_witnesses.add_vkey_key_hash(hash);
+            },
+            StakeCredential::Script { hash, .. } => {
+                required_witnesses.add_script_hash(hash);
+                // TODO: my understand is that inline datums in inputs also need to be added to the witness
+                // but I still need to confirm this by actually submitting a transaction that tests this
+                // see same comment in add_reference_input
+                if let Some(data_hash) = utxo_info.datum_hash() {
+                    required_witnesses.add_plutus_datum_hash(data_hash);
+                    // note: redeemer is required as well
+                    // but we can't know the index, so we rely on the tx builder to satisfy this requirement
+                }
+            },
+        }
+    };
+    if let Address::Byron(byron) = utxo_info.address() {
+        required_witnesses.add_bootstrap(byron);
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct InputBuilderResult {
+    pub(crate) input: TransactionInput,
```

Comment:
Why pub(crate)? there's no way to access fields externally when you get this result, it's not very convenient